core/lib/spree/event.rb

```diff
@@ -6,44 +6,102 @@
 require_relative 'event/subscriber'
 
 module Spree
+  # Event bus for Solidus.
+  #
+  # This module serves as the interface to access the Event Bus system in
+  # Solidus. You can use different underlying adapters to provide the core
+  # logic. It's recommended that you use {Spree::Event::Adapters::EventBus}.
+  #
+  # You use the {#fire} method to trigger an event:
+  #
+  #  @example
+  #    Spree::Event.fire 'order_finalized', order: order
+  #
+  # Then, you can use {#subscribe} to hook into the event:
+  #
+  #   @example
+  #     Spree::Event.subscribe 'order_finalized' do |event|
+  #       # Take the order at event.payload[:order]
+  #     end
+  #
+  # You can also subscribe to an event through a module including
+  # {Spree::Event::Subscriber}:
+  #
+  #  @example
+  #    module MySubscriber
+  #      include Spree::Event::Subscriber
+  #
+  #      event_action :order_finalized
+  #
+  #      def order_finalized(event)
+  #        # Again, take the order at event.payload[:order]
+  #      end
+  #    end
   module Event
     extend self
 
     delegate :activate_autoloadable_subscribers, :activate_all_subscribers, :deactivate_all_subscribers, to: :subscriber_registry
 
-    # Allows to trigger events that can be subscribed using #subscribe. An
-    # optional block can be passed that will be executed immediately. The
-    # actual code implementation is delegated to the adapter.
+    # Allows to trigger events that can be subscribed using {#subscribe}.
     #
-    # @param [String] event_name the name of the event. The suffix ".spree"
-    #  will be added automatically if not present
-    # @param [Hash] opts a list of options to be passed to the triggered event
+    # The actual code implementation is delegated to the adapter.
+    #
+    # @param [String, Symbol] event_name the name of the event.
+    # @param [Hash] opts a list of options to be passed to the triggered event.
+    # They will be made available through the {Spree::Event::Event} instance
+    # that is yielded to the subscribers (see {Spree::Event::Event#payload}).
+    # However, take into account that the deprecated
+    # {Spree::Event::Adapters::ActiveSupportNotifications} adapter will yield a
+    # {ActiveSupport::Notifications::Fanout::Subscribers::Timed} instance
+    # instead.
+    # @option opts [Any] :adapter Reserved to indicate the adapter to use as
+    # event bus. Defaults to {#default_adapter}
     #
     # @example Trigger an event named 'order_finalized'
     #   Spree::Event.fire 'order_finalized', order: @order do
     #     @order.finalize!
     #   end
+    #
+    # TODO: Change opts to be keyword arguments and include `adapter:` in them
```

Comment:
Is this `TODO` for this PR?

Reply:
I'd say no, but let me explain it better.

So, the issue is that I'd like to be consistent with the other methods, as I do think that having `adapter:` as a keyword argument is the most expressive way to inject adapters. However, if I add it to the signature:

```ruby
def fire(event_name, opts = {}, adapter: default_adapter)
```

There's no way for it to work on Ruby < 3 unless we provide the `adapter:` argument when invoking it (which breaks the whole purpose). I.e., both

```ruby
fire('foo', order: order)
```

and

```ruby
fire('foo', { order: order })
```

will raise `ArgumentError (unknown keyword: :order)`.

Given that, we should move `opts` from being a `Hash` to keyword arguments:

```ruby
def fire(event_name, adapter: default_adapter, **kwargs)
```

It will work in RUBY < 3, both with:

```ruby
fire('foo', order: order)
```

and

```ruby
fire('foo', { order: order })
```

Still, Ruby 3 stores are now expecting that the event options are given as a `Hash`, so it will break them. In addition, we can't provide a straightforward deprecation path as if we try to support both options:

```ruby
def fire(event_name, opts = {}, adapter: default_adapter, **kwargs)
```

we're back to square one.

I can think of two workarounds for it:

1. Define different methods depending on `RUBY_VERSION`. E.g. (pseudo-code):

```ruby
if RUBY_VERSION < 3
  def fire(event_name, adapter: default_adapter, **kwargs)
    # ...
else
  def fire(event_name, opts = {}, adapter: default_adapter, **kwargs)
    opts = if opts.any?
                deprecate('Use kwargs instead of hash')
                opts
              else
                kwargs
              end
```

I think this option would deserve a separate commit.

2. Just change the signature to:

```ruby
def fire(event_name, adapter: default_adapter, **kwargs)
```

on Solidus 4 and inform about the change on the upgrade guide.

Reply:
Got it. What if we use opts for now and deprecate all the un-recognized options passed? So that when we'll switch to kwargs in Solidus 4 we'll sure to be compliant.



```ruby
# Pseudo-code

def fire(event_name, opts = {}, 
  adapter = opts.delete :adapter
  order = opts.delete :order
  
  opts.each do |key, value|
     Spree::Deprecation.warn("passing #{key} to fire is deprecated, please just pass order and adapter")
  end
end
```

Reply:
The problem is that when you fire an event, you can attach to it whatever information you want to make available to the subscribers. Here I used `order`, but it could be anything else :slightly_smiling_face:

Reply:
Oh right, how did I miss it? I see no way to deprecate it before removal then, maybe just changing the method name and deprecate `fire` at all but I don't think it's worth it. We can specify in the TODO that we can remove it in the next major, adding the keyword Deprecation somehow so that we find it along with the others when removing 3.x deprecations. 

Alternatively we could create a new issue for this, and move it into a new 4.0 milestone in GitHub.

Reply:
I added the improved `TODO:` comment with `Spree::Deprecation` on it a link to this comment :slightly_smiling_face: